### PR TITLE
ObjCMessageTracing: append trace file content atomically to a canonical file location provided by the build environment

### PR DESCRIPTION
**Pull request by nkcsgexi**
To simplify post-processing of these collected ObjC message traces, this PR teaches the compiler to atomically append the content of ObjC message trace to a file location provided by the build environment, instead of emitting multiple files.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Review by DougGregor — COMMENTED** [review]
IIRC, the module trace format is fairly extensible. Should we extend that file rather than threading another trace file through the driver and frontend?

**Comment by nkcsgexi** [comment]
> IIRC, the module trace format is fairly extensible. Should we extend that file rather than threading another trace file through the driver and frontend?

We might be just using the existing module trace file for these fine-grained symbol dependencies as well, yeah. However, we've heard one motivation for keeping them separate so that we could to keep each of these trace file relatively small for downloading/post-processing.

**Comment by nkcsgexi** [comment]
That's a reasonable suggestion. Will reposition this file to be a more general-purposed one!

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Frontend/ArgsToFrontendOutputsConverter.cpp [hunk 1]
```diff
@@ -455,6 +455,24 @@ SupplementaryOutputPathsComputer::computeOutputPathsForOneInput(
...
       defaultSupplementaryOutputPathExcludingExtension);
 
+  // We piggy-back on the loadedModuleTracePath to decide (1) whether
+  // to emit the fine module Trace file, and (2) where to emit the fine module
+  // trace file if the path isn't explicitly given by
+  // SWIFT_COMPILER_OBJC_MESSAGE_TRACE_PATH.
+  // FIXME: we probably need to move this to a frontend argument.
+  llvm::SmallString<128> FineModuleTracePath;
+  if (!loadedModuleTracePath.empty()) {
+    if (const char *P = ::getenv("SWIFT_COMPILER_OBJC_MESSAGE_TRACE_PATH")) {
```
Comment (DougGregor) [inline lib/Frontend/ArgsToFrontendOutputsConverter.cpp:465]: Can you remove the OBJC from this and use FINE_GRAINED instead?

Comment (nkcsgexi) [reply lib/Frontend/ArgsToFrontendOutputsConverter.cpp:465]: Will fix!

**Pull request merged**